app/uk/gov/hmrc/hec/services/scheduleService/HecTaxCheckExtractionService.scala

```diff
@@ -129,50 +125,50 @@ class HecTaxCheckExtractionServiceImpl @Inject() (
   private def toFormattedString(i: Int) = f"$i%04d"
 
   private def createHecFile(
-    hecTaxCheckList: HECTaxCheckFileBodyList,
-    count: Int,
+    limit: Int,
     partialFileName: String,
     dirname: String
-  ): EitherT[Future, Error, List[Unit]] = {
-    @tailrec
-    def loopHecTaxCheckRecords(
-      hecTaxCheckList: List[HECTaxCheck],
-      count: Int,
+  ): EitherT[Future, Error, Unit] = {
+
+    @SuppressWarnings(Array("org.wartremover.warts.All"))
+    def loop(
       seqNumInt: Int,
+      skip: Int,
+      limit: Int,
+      sortBy: String,
       partialFileName: String,
-      dirname: String,
-      acc: List[EitherT[Future, Error, Unit]]
-    ): List[EitherT[Future, Error, Unit]] = {
-      val hecList          = hecTaxCheckList.take(count)
-      val remainingList    = hecTaxCheckList.drop(count)
-      val isLastInSequence = remainingList.size === 0
-      //If list is empty and seq number is not 1 or seq num > 9999, halt the process
-      if ((hecList.size === 0 && seqNumInt =!= 1) || seqNumInt > 9999) {
-        logger.info(
-          "Hec tax Check file creation process halted:: Either there are no more records to process or the sequence number exceeds 9999. "
-        )
-        acc
-      } else {
-        loopHecTaxCheckRecords(
-          remainingList,
-          count,
-          seqNumInt + 1,
-          partialFileName,
-          dirname,
-          createAndStoreFileThenNotify(
-            HECTaxCheckFileBodyList(hecList),
-            toFormattedString(seqNumInt),
-            partialFileName,
-            dirname,
-            isLastInSequence
-          ) :: acc
-        )
+      dirname: String
+    ): EitherT[Future, Error, Unit] = {
+
+      val fetchNextBatchHECTaxCheckData = for {
+        hecTaxCheckList          <- taxCheckService.getAllTaxCheckCodesByExtractedStatus(false, skip, limit, sortBy)
+        _                         =
+          logger.info(
+            s" Processing file no :: $seqNumInt, with records:: ${hecTaxCheckList.size}}"
+          )
+        hecTaxCheckListNextBatch <-
```

Comment:
it seems like we may be doing more calls to get on the database than necessary. If there's 8 things in the database and the limit is 5 for instance I think we'd have:
```scala
// loop 1
hecTaxCheckList <- taxCheckService.getAllTaxCheckCodesByExtractedStatus(false, 0, 5, sortBy) // 5 things
hecTaxCheckListNextBatch <- taxCheckService.getAllTaxCheckCodesByExtractedStatus(false, 5, 5, sortBy) // 3 things

// loop 2
hecTaxCheckList <- taxCheckService.getAllTaxCheckCodesByExtractedStatus(false, 5, 5, sortBy) // 3 things
hecTaxCheckListNextBatch <- taxCheckService.getAllTaxCheckCodesByExtractedStatus(false, 10, 5, sortBy) // 0 things
```
so it looks like there'd be two calls with
```
taxCheckService.getAllTaxCheckCodesByExtractedStatus(false, 5, 5, sortBy) 
```
If you have a case like that in the spec where it loops through more than once you'd see it I think. To avoid that we could maybe pass in the "currentBatch" as a parameter in the loop so that we only call it once per page.

Also, having written the above there's a condition that we could look out for to save some mongo calls - if the current page has strictly fewer records than the limit then we don't need to make a call to get the next page since we know we've got to the end of the whole batch. So I think we could have something like:
```scala
hecTaxCheckListNextBatch <- if(currentBatch.size < limit)  EitherT.pure(List.empty)
                            else taxCheckService.getAllTaxCheckCodesByExtractedStatus(false, skip + limit, limit, sortBy)
```

Reply:
yeah , i have scenario like that  "taxCheckService.getAllTaxCheckCodesByExtractedStatus(false, 5, 5, sortBy)  in spec. I can try to work on that and second one as well

Reply:
hat if i move the hecTaxCheckListNextBatch call in else part, then in that case it will be called only when the current batch has data

Reply:
yeah, in that case, the if in flatmap will not make sense 🤦‍♀️

Reply:
I have done something like this
def loop(same param)={
val record = for{
firstbatch <-db call
_ <- check if its' size is 0, then return 0, else  call second batch ,do create file processing  return the count  of 2nd batch
}

then record .flatMap{same}
In this case we will be fetching second batch  only when needed
what do you think?

Reply:
wouldn't it still do the duplicate calls in the intermediate loops? I was thining of something like this:
```scala
private def createHecFile(
    limit: Int,
    partialFileName: String,
    dirname: String
  ): EitherT[Future, Error, Unit] = {

    @SuppressWarnings(Array("org.wartremover.warts.All"))
    def loop(
      seqNumInt: Int,
      skip: Int,
      limit: Int,
      sortBy: String,
      partialFileName: String,
      dirname: String,
      currentBatch: EitherT[Future, Error, Unit]
    ): EitherT[Future, Error, Unit] = {

      val fetchNextBatchHECTaxCheckData = for {
        hecTaxCheckList          <- currentBatch
        _                         =
          logger.info(
            s" Processing file no :: $seqNumInt, with records:: ${hecTaxCheckList.size}}"
          )
        hecTaxCheckListNextBatch <-
          taxCheckService.getAllTaxCheckCodesByExtractedStatus(false, skip + limit, limit, sortBy)
        _                        <- if ((hecTaxCheckList.size === 0 && seqNumInt =!= 1) || seqNumInt > 9999) EitherT.pure[Future, Error](())
                                    else {
                                      createAndStoreFileThenNotify(
                                        HECTaxCheckFileBodyList(hecTaxCheckList),
                                        toFormattedString(seqNumInt),
                                        partialFileName,
                                        dirname,
                                        hecTaxCheckListNextBatch.size === 0
                                      )
                                    }
      } yield hecTaxCheckListNextBatch

      fetchNextBatchHECTaxCheckData.flatMap { hecTaxCheckList =>
        if (hecTaxCheckList.size === 0) EitherT.pure[Future, Error](())
        else loop(seqNumInt + 1, skip + limit, limit, sortBy, partialFileName, dirname, fetchNextBatchHECTaxCheckData)
      }
    }

    loop(
        1,
        0,
        limit,
        "_id",
        partialFileName,
        dirname,
        taxCheckService.getAllTaxCheckCodesByExtractedStatus(false, 0, limit, sortBy))
  }
```
Though it is slightly annoying to `flatMap`into `fetchNextBatchHECTaxCheckData` and pass it into the loop again

Reply:
in this case also , we are doing it twice, one in the  loop param  and inside hecTaxCheckListNextBatch <-
          taxCheckService.getAllTaxCheckCodesByExtractedStatus(false, skip + limit, limit, sortBy) . Let me try this one and will see in test case

Reply:
I have updated the code as per your suggestion. It's working , added test cases to show that